dgisim/tests/test_dices.py

```diff
@@ -273,3 +274,339 @@ def test_ordered_actual_dices_with_characters(self):
+                        local_precedence_arg=input_precedence
+                    )
+
+                    expected = None
+                    self.assertEqual(actual, expected)
+
+    def test_smart_selection_code_cover(self):
+        """
+        Element and OMNI feed OMNI
+        code cover
+        """
+        first_element = Element.GEO  # most precedent element
+
+        for first_element_supply_number in range(5):
+            for omni_supply_number in range(5):
+                for omni_need_number in [1, 3, 5]:
+                    with self.subTest(first_element=first_element,
+                                      first_element_supply_number=first_element_supply_number,
+                                      omni_supply_number=omni_supply_number, omni_need_number=omni_need_number):
+                        first_element_supply_number = 3
+                        omni_supply_number = 2
+                        omni_need_number = 4
+                        input_actual_dices: ActualDices = ActualDices(
+                            {first_element: first_element_supply_number, Element.OMNI: omni_supply_number})
+                        input_abstract_dices: AbstractDices = AbstractDices(
+                            {Element.OMNI: omni_need_number})
+                        input_precedence: list[set[Element]] = [{first_element}]
+
+                        actual = input_actual_dices.smart_selection(
+                            requirement=input_abstract_dices,
+                            game_state=None,
+                            local_precedence_arg=input_precedence
+                        )
+
+                        if first_element_supply_number + omni_supply_number >= omni_need_number:
+                            expected = ActualDices({first_element: first_element_supply_number,
+                                                    Element.OMNI: omni_need_number - first_element_supply_number})
+                        else:
+                            expected = None
+                        self.assertEqual(actual, expected)
+
+    def test_smart_selection_negative_2(self):
+        """
+        test for code coverage.
+        supply: 2 of every elements
+        need: 5 omni
+        """
+        elements_and_omni = (Element.OMNI,) + ActualDices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+        input_actual_dices: ActualDices = ActualDices(
+            dices={el: 2 for el in elements_and_omni}
+        )
+        input_abstract_dices: AbstractDices = AbstractDices(
+            dices={Element.OMNI: 5}
+        )
+        input_precedence: list[set[Element]] = [
+            {el} for el in ActualDices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+        ]
+
+        actual: Optional[ActualDices] = input_actual_dices.smart_selection(
+            requirement=input_abstract_dices, game_state=None, local_precedence_arg=input_precedence
+        )
+
+        expected = None
+        self.assertEqual(actual, expected)
```

Comment:
unrelated to this line of code, but `smart_selection()` fails the following test (I coded):

```py
def test_lesser_omni_spent_when_two_first_priority_elems_present(self):
    """
    Test if the better less-omni-costing plan is chosen when there are two first priority
    elements present with different numbers.
    """
    for n1, n2 in ((1, 2), (2, 1)):
        with self.subTest(n1=n1, n2=n2):
            requirement = AbstractDices({Element.OMNI: 3})
            dices = ActualDices({Element.OMNI: 2, Element.CRYO: n1, Element.ANEMO: n2})
            selected_dices = dices.smart_selection(requirement, None, local_precedence_arg=[{Element.CRYO, Element.ANEMO}])
            self.assertEqual(selected_dices[Element.OMNI], 1)
```

Reply:
fixed.